testcases/OpTestKexec.py

```diff
@@ -0,0 +1,129 @@
+import OpTestLogger
+
+import OpTestConfiguration
+from common.OpTestSystem import OpSystemState
+from common.Exceptions import CommandFailed
+from common.OpTestError import OpTestError
+
+log = OpTestLogger.optest_logger_glob.get_logger(__name__)
+
+class OpTestKexec(unittest.TestCase):
+
+    def setUp(self):
+        conf = OpTestConfiguration.conf
+        self.cv_SYSTEM = conf.system()
+        self.c = self.cv_SYSTEM.console
+        self.kernel_image = conf.args.kernel_image
+        self.initrd_image = conf.args.initrd_image
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+        res = self.c.run_command("cat /etc/os-release")
+        if "Ubuntu" in (res[0] or res[1]):
+            self.distro = "ubuntu"
+        elif 'Red Hat' in (res[0] or res[1]):
+            self.distro = 'rhel'
+        elif 'SLES' in (res[0] or res[1]):
+            self.distro = 'sles'
+        else:
+            raise self.skipTest("Test currently supported only on ubuntu, sles and rhel")
+
+ 
+    def kexec_load(self):
+        """
+        This function will test the kexec load
+        """
+
+        try:
+            self.c.run_command("kexec -l /boot/{} --initrd /boot/{} --append=\"`cat /proc/cmdline`\"".format(self.kernel_image,self.initrd_image))
+        except CommandFailed:
+            kexec_loaded_status = self.c.run_command("cat /sys/kernel/kexec_loaded")
+            if kexec_loaded_status[0] == "0":
+                log.info("New kernel not loaded- Failed with kexec_loaded_status = \"%s\"", kexec_loaded_status[0])
+            raise OpTestError("kexec -l failed to load kernel")
+        
+    def kexec_unload(self):
+        """
+        This function will test the kexec unload
+        """
+
+        try:
+            self.c.run_command("kexec -u")
+        except CommandFailed:
+            kexec_loaded_status = self.c.run_command("cat /sys/kernel/kexec_loaded")
+            if kexec_loaded_status[0] == "1":
+                log.info("kernel unload Failed with kexec_loaded_status = \"%s\"", kexec_loaded_status[0])
+            raise OpTestError("Kexec -u  failed to unload")
+        
+    def kexec_load_reboot(self):
+        """
+        This function will test the kexec load and reboot
+        """
+
+        try:
+            self.c.run_command("kexec -l /boot/{} --initrd /boot/{} --append=\"`cat /proc/cmdline`\"".format(self.kernel_image,self.initrd_image))
+        except CommandFailed:
+            kexec_loaded_status = self.c.run_command("cat /sys/kernel/kexec_loaded")
+            if kexec_loaded_status[0] == "0":
+                log.info("New kernel not loaded- Failed with kexec_loaded_status = \"%s\"", kexec_loaded_status[0])
+            raise OpTestError("kexec -l failed to load kernel")
+
+        try:
+            self.c.run_command("kexec -e")
+        except CommandFailed:
+            log.info("System is booting. Let's wait")
+            time.sleep(240)
+            self.cv_SYSTEM.goto_state(OpSystemState.OS)
```

Comment:
I don't think this is the correct way. We should also capture the console messages during kexec boot. Without them there is no way know why the kernel boot failed. If you capture the console then there is no need to use sleep() function.

Reply:
Used raw_pty to get console and send line . Please check output format copied below :
[console-expect]#kexec -l /boot/vmlinux-5.3.18-57-default --initrd /boot/initrd-5.3.18-57-default --append="`cat /proc/cmdline`"
kexec -l /boot/vmlinux-5.3.18-57-default --initrd /boot/initrd-5.3.18-57-default --append="`cat /proc/cmdline`"
Modified cmdline:BOOT_IMAGE=/boot/vmlinux-5.3.18-57-default root=UUID=4e8439c7-7775-4160-950d-41aee0bed249 splash=verbose mitigations=auto quiet fadump= crashkernel=242M
[console-expect]#echo $?
echo $?
0
[console-expect]#kexec -e
kexec -e
[  912.330905] kexec_core: Starting new kernel
I'm in purgatory
.
.
.
.
[  OK  ] Started OpenSSH Daemon.
         Starting Locale Service...
[  OK  ] Started Locale Service.


Welcome to SUSE Linux Enterprise Server 15 SP3  (ppc64le) - Kernel 5.3.18-57-default (hvc0).

eth0: 9.40.194.147 2002:903:15f:550:796b:b48e:da3b:d91d


localhost login: ok

----------------------------------------------------------------------
Ran 1 test in 28.472s